### PR TITLE
Add index to Analysis Server documentation

### LINKED ISSUE #23971: analysis server spec html generator: add index for easier navigation of html file
This was discussed a while ago in the analyzer mailing list. I can have a go at implementing an index that links to the requests, types etc from the bottom of the file. The api has grown quite a bit and hunting down the item you're interested in looking up is no fun. Hopefully, an index would make it simpler to browse the different types, requests, etc.

Very rough WIP below.

index

Suggestions welcome! I'm planning to use tables to compact the sections, very much like the Python docs do for basically the same thing; see: https://docs.python.org/3/library/functions.html

### PR DESCRIPTION
**Pull request by guillermooo**
Fixes #23971 

I've backed out from using a two-col layout for displaying subindex items. This perhaps be implemented later if necessary.

Here's a preview of the output:

http://htmlpreview.github.io/?https://gist.githubusercontent.com/guillermooo/0f90ceea066c0d3d8d0f/raw/5500136eafeecfbce7dc88f48b61a53054a82e3e/dart_das_api_doc.html

I think his PR is ready now.

@bwilkerson

### CODE REVIEW

**Comment by bwilkerson** [comment]
It looks good to me. Let me know when you think it's done.

**Comment by guillermooo** [comment]
I think this is done now. Updated preview here:

http://htmlpreview.github.io/?https://gist.githubusercontent.com/guillermooo/0f90ceea066c0d3d8d0f/raw/8e73ef76e60e4605272a3cc6cf45d3e98e646157/dart_das_api_doc.html

Note the line about the index under "Domains" at the top of the file, and the link to the index from there.

**Comment by bwilkerson** [comment]
It looks good! As a final step, could you run the generator and add any modified files to this PR?

As you might know, we have a step in the build process that runs the generator and compares the resulting files against what's been committed in order to catch potential problems when the generated code hasn't been updated to match the spec. The generated code shouldn't change because of these changes, but the generated HTML doc will, so our build would break at this point.

**Comment by guillermooo** [comment]
I get some issues related to Windows, I suppose.

```
dart -c generate_all.dart
```

With `core.autocrfl` set to `true` in git, the `generated/**` files are marked as 'modified' by git. After diffing, it seems to be due to newline chars. In fact, if I disable `auto.corecrlf` temporarily and do this:

```
git rm generated/*
git reset --hard
dart -c generate_all.dart
```

Then now I only get these modified files:

```
        modified:   ../../doc/api.html
        modified:   ../../lib/src/generated_protocol.dart
        modified:   ../../test/integration/integration_test_methods.dart
        modified:   ../../test/integration/protocol_matchers.dart
```

And of those, all but the first get marked due to newline chars again. (I've tried checking those out with LF on Windows and re-running the generate_all.dart script, but got the same results. I've probvably done something wrong or git simply doesn't like Windows as usual.)

#### Short version

Only api.html includes meaningful changes, as expected, so I'm committing that only shorly.

**Comment by guillermooo** [comment]
OK, so the problem seems to be that the Dart generator uses LF on Windows too and git gets upset on Windows *depending on your EOL configuration*.

It seems that the generator should use CRLF on Windows to avoid such problems when using the recommended (by some) config: `core.autocrlf = true`.

Note that if you have different settings globally and local to the repo, weird things happen. This is likely all git EOL spelunking I'll do for this issue, but the conclusion is that the generated files, as expected, are identical except for `api.html`.

**Comment by guillermooo** [comment]
This whole sequence (Windows) shows that there are in fact no actual changes:

```
spec eol-test $ git rm ../../*
spec eol-test* $ git reset --hard
spec eol-test $ dart -c .\generate_all.dart
spec eol-test* $ git st
On branch eol-test
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   ../../doc/api.html
        modified:   ../../lib/src/generated_protocol.dart
        modified:   ../../test/integration/integration_test_methods.dart
        modified:   ../../test/integration/protocol_matchers.dart
        modified:   generated/java/AnalysisServer.java
       [ ... everything under `generated` follows ]

spec eol-test* $ git add ../..
warning: LF will be replaced by CRLF in pkg/analysis_server/doc/api.html.
The file will have its original line endings in your working directory.
warning: LF will be replaced by CRLF in pkg/analysis_server/lib/src/generated_protocol.dart.
The file will have its original line endings in your working directory.
warning: LF will be replaced by CRLF in pkg/analysis_server/test/integration/integration_test_methods.dart.
The file will have its original line endings in your working directory.
warning: LF will be replaced by CRLF in pkg/analysis_server/test/integration/protocol_matchers.dart.
The file will have its original line endings in your working directory.
warning: LF will be replaced by CRLF in pkg/analysis_server/tool/spec/generated/java/AnalysisServer.java.
The file will have its original line endings in your working directory.
[ ... everything under `generated` follows ]

spec eol-test $ git st
On branch eol-test
nothing to commit, working directory clean  
```

**Comment by guillermooo** [comment]
@bwilkerson 

BTW, I've just tried to install `depot_tool` and it seems to have worked (last time I tried, I got an error). I f necessary, I can try to follow the standard procedure to submit a patch.

**Comment by bwilkerson** [comment]
> ... I can try to follow the standard procedure to submit a patch.

No need. Sorry for the delay, I ended up taking Friday off.
